crates/bevy_ecs/src/query/iter.rs

```diff
@@ -220,6 +856,151 @@ impl<'w, 's, D: QueryData, F: QueryFilter> Iterator for QueryIter<'w, 's, D, F>
+where
+    I: Iterator<Item = Entity>,
+{
+    /// # Safety
+    /// - `world` must have permission to access any of the components registered in `query_state`.
+    /// - `world` must be the same one used to initialize `query_state`.
+    /// - `entity_list` must only contain unique entities or be empty.
+    pub(crate) unsafe fn new<EntityList: IntoIterator<IntoIter = I>>(
+        world: UnsafeWorldCell<'w>,
+        query_state: &'s QueryState<D, F>,
+        entity_list: EntityList,
+        last_run: Tick,
+        this_run: Tick,
+    ) -> QuerySortedIter<'w, 's, D, F, I> {
+        let fetch = D::init_fetch(world, &query_state.fetch_state, last_run, this_run);
+        let filter = F::init_fetch(world, &query_state.filter_state, last_run, this_run);
+        QuerySortedIter {
+            query_state,
+            entities: world.entities(),
+            archetypes: world.archetypes(),
+            // SAFETY: We only access table data that has been registered in `query_state`.
+            // This means `world` has permission to access the data we use.
+            tables: &world.storages().tables,
+            fetch,
+            filter,
+            entity_iter: entity_list.into_iter(),
+        }
+    }
+
+    /// # Safety
+    /// `entity` must stem from `self.entity_iter`, and not have been passed before.
+    #[inline(always)]
+    unsafe fn fetch_next(&mut self, entity: Entity) -> Option<D::Item<'w>> {
+        let (location, archetype, table);
+        // SAFETY:
+        // `tables` and `archetypes` belong to the same world that the [`QueryIter`]
+        // was initialized for.
+        unsafe {
+            location = self.entities.get(entity).debug_checked_unwrap();
+            archetype = self
+                .archetypes
+                .get(location.archetype_id)
+                .debug_checked_unwrap();
+            table = self.tables.get(location.table_id).debug_checked_unwrap();
+        }
+
+        // SAFETY: `archetype` is from the world that `fetch/filter` were created for,
+        // `fetch_state`/`filter_state` are the states that `fetch/filter` were initialized with
+        unsafe {
+            D::set_archetype(
+                &mut self.fetch,
+                &self.query_state.fetch_state,
+                archetype,
+                table,
+            );
+        }
+        // SAFETY: `table` is from the world that `fetch/filter` were created for,
+        // `fetch_state`/`filter_state` are the states that `fetch/filter` were initialized with
+        unsafe {
+            F::set_archetype(
+                &mut self.filter,
+                &self.query_state.filter_state,
+                archetype,
+                table,
+            );
+        }
+
+        // SAFETY: set_archetype was called prior.
+        // `location.archetype_row` is an archetype index row in range of the current archetype.
+        if unsafe { F::filter_fetch(&mut self.filter, entity, location.table_row) } {
```

Comment:
F::filter_fetch could be removed as we can assume that all relevant entities have already undergone filtering.

Reply:
good catch!